docs/testing-with-django.md

```diff
@@ -1,8 +1,237 @@
+tests that might not consider ancillary parts of data setup that could
+influence tests results.)
+
+Here are some example action methods that tests may use for data setup:
+
+- check_send_message
+- create_stream_if_needed
+- do_add_subscription
+- do_change_is_admin
+- do_create_user
+- do_make_stream_private
+
+## Zulip Testing Philosophy
+
+If there is one word to describe Zulip's philosophy for writing tests,
+it is probably "flexible."  (Hopefully "thorough" goes without saying.)
+
+When in doubt, unless speed concerns are prohibitive,
+you usually want your tests to be somewhat end-to-end, particularly
+for testing endpoints.
+
+These are some of the testing strategies that you will see in the Zulip
+test suite...
+
+### Endpoint tests
+
+We strive to test all of our URL endpoints.  The vast majority of Zulip
+endpoints support a JSON interface.  Regardless of the interface, an
+endpoint test generally follows this pattern:
+
+- Set up the data.
+- Login with `self.login()` or set up an API key.
+- Use a Zulip test helper to hit the endpoint.
+- Assert that the result was either a success or failure.
+- Check the data that comes back from the endpoint.
+
+Generally, if you are doing endpoint tests, you will want to create a
+test class that is a subclass of `AuthedTestCase`, which will provide
+you helper methods like the following:
+
+- api_auth
+- assert_json_error
+- assert_json_success
+- client_get
+- client_post
+- get_api_key
+- get_streams
+- login
+- send_message
+
+### Library tests
+
+For certain Zulip library functions, especially the ones that are
+not intrinsically tied to Django, we use a classic unit testing
+appproach of calling the function and inspecting the results.
+
+For these types of tests, you will often use methods like
+`self.assertEqual()`, `self.assertTrue()`, etc., which come with
+Django.
+
+### Fixture-driven tests
+
+Particularly for testing Zulip's integrations with third party systems,
+we strive to have a highly data-driven approach to testing.  To give a
+specific example, when we test our GitHub integration, the test code
+reads a bunch of sample inputs from a JSON fixture file, feeds them
+to our Github integration code, and then verifies the output against
+expected values from the same JSON fixture file.
+
+Our fixtures live in `zerver/fixtures`. 
+
+### Mocks and stubs
+
+We use mocks and stubs for all the typical reasons:
+
+- to more precisely test the target code
+- to stub out calls to third-party services
+- to make it so that you can run your tests on the airplane without wifi
+
+For mocking we generally use the "mock" library and use `mock.patch` as
+a context manager.  We also take advantage of some context managers
+from Django as well as our own custom helpers.  Here is an example:
+
+
+        with self.settings(RATE_LIMITING=True):
+            with mock.patch('zerver.decorator.rate_limit_user') as rate_limit_mock:
+                api_result = my_webhook(request)
```

Comment:
For settings in particular, we should link to this content: http://zulip.readthedocs.io/en/latest/settings.html#testing-non-default-settings